man/getrec.3
.TH GETREC 3 " 4 April 1996"
.FB
.SH NAME
getrec
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_getrec(rec, db)
.br
long rec;
.br
fb_database *db;
.br
.PP
.SH DESCRIPTION
As explained in the
\fIdatabase\fP(4) manual page, a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
\fIFb_getrec\fP
retrieves logical record \fIrec\fP from the database pointed to by \fIdb\fP
and initializes the array of field structures.
The map file descriptor stored in \fIdb->mfd\fP
is used to determine the location and length of the
record in the database. The database file descriptor,
stored in \fIdb->fd\fP, is used to read the indicated record.
.PP
As mentioned,
\fIfb_getrec\fP sets up all of the database field descriptors so each
individual field is readily available. Each field is accessed through
an array of field descriptors, or key pointers. The base of this array
is located in the database structure variable \fIdb->kp\fP.
.PP
A macro function, \fBFB_FLD(n, db)\fP can be used to access
field number \fIn\fP of database \fIdb\fP, or use \fIfetch\fP(3).
.SH CLIENT/SERVER
The \fIfb_getrec\fP routine will transparently
call the \fIfb_getrec_clnt\fP mechanism
when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP is set to one.
See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
fetch(3), putrec(3), getirec(3), getxrec(3), opendb(3), database(4).
.SH DIAGNOSTICS
\fIFb_getrec\fP will return the error code FB_ERROR if any of its reads,
writes, or seeks fail. On correct completion, the number of characters read
will be returned.
.PP
Note: \fIgetrec\fP will lock record 0 to gain exclusive access to
the database only if the LOCKLEVEL is \fB2\fP or more. See \fIsetup\fP(5).
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/getrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: getrec.c,v 9.0 2001/01/09 02:56:58 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Getrec_sid[] = "@(#) $Id: getrec.c,v 9.0 2001/01/09 02:56:58 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

static long min_rlen = FB_MAPREC_SIZE + 2;
extern short int cdb_allow_links;
extern short int cdb_use_rpc;
extern short int cdb_locklevel;
extern short int cdb_dbase_byteorder;
extern short int cdb_cpu_byteorder;

#if FB_PROTOTYPES
static sread(fb_database *hp, long rlen);
#else
static sread();
#endif /* FB_PROTOTYPES */

extern char cdb_EOREC;

/* 
 * getrec - get logical record n from fd into hp->orec using hp->mfd as map.
 *	this map provides the position and the length of the record.
 * 	the fb_field pointers are loaded into hp->kp.
 * 	the Actual number of characters in the record is returned.
 *	note: the assumption is that sizeof(int) is plenty for max reclen,
 *	-- thus comparison of int i to long rlen in sread().
 */

   fb_getrec(n, hp)
      long n;
      register fb_database *hp;
   
      {
         long rpos, rlen, fend;
         char buf[FB_SLONG+FB_SLONG+3];
	 int status;
         int map_rlen = FB_SLONG + FB_SLONG;

#if RPC
         if (cdb_use_rpc)
            return(fb_getrec_clnt(n, hp));
#endif /* RPC */
	 hp->rec = n;			/* store this rec num */
	 
	 /* 
	  *  seek past map record n since first record is header, meaning
	  *  the point of interest is past the n'th physical record.
	  */

	 if (cdb_locklevel > 1)
	    fb_lock_head(hp);
         if (lseek(hp->mfd, n * (long) FB_MAPREC_SIZE, 0) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
	    return(FB_ERROR);
	    }
	 if (read(hp->mfd, buf, (unsigned) map_rlen) != map_rlen){
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
	    return(FB_ERROR);
	    }
         memcpy((char *) &rpos, buf, FB_SLONG);
         memcpy((char *) &rlen, buf + FB_SLONG, FB_SLONG);
         if (cdb_dbase_byteorder != cdb_cpu_byteorder){
            M_32_SWAP(rpos);
            M_32_SWAP(rlen);
            }

	 if (rlen < min_rlen || 
	       (rlen > hp->recsiz + 1 && hp->recsiz > min_rlen)){
	    fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
	    return(FB_ERROR);
	    }
	 if (lseek(hp->fd, rpos, 0) < 0L){
	    fb_lerror(FB_SEEK_ERROR, SYSMSG[S_BAD_DATA], hp->dbase);
	    return(FB_ERROR);
	    }
	 if ((status = sread(hp, rlen)) == FB_ERROR){
            /* test rpos against end of file. if >, its a map prob */
	    fend = lseek(hp->fd, 0L, 2);
            if (rpos >= fend)
	       fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
            else
	       fb_lerror(FB_READ_ERROR, SYSMSG[S_BAD_DATA], hp->dbase);
	    return(FB_ERROR);
	    }
	 if (cdb_locklevel > 1)
	    fb_unlock_head(hp);
         if (cdb_allow_links)
	    fb_getlink(hp);
	 return(status);
      }

/*
 *  sread - do a simple read up to rlen characters or eorec char.
 *	return number of chars actually in the record.
 */
 
   static sread(hp, rlen)
      fb_database *hp;
      long rlen;
   
      {
         register int i;
	 register char *buf, *p;
	 fb_field **kf, *lf;
	 
	 FB_FLD(0,hp) = buf = hp->orec;		/* set FB_FLD 0 and buf */
	 if (read(hp->fd, buf, (unsigned) rlen) != rlen)
	    return(FB_ERROR);
	 i = 1, kf = hp->kp, lf = hp->kp[hp->nfields];
	 for (; ; buf++, i++){
            if ((*kf)->type == FB_BINARY){
               buf += ((*kf)->size);
               i += ((*kf)->size);
               if (*buf == cdb_EOREC)		/* could be last fb_field */
                  break;
               buf--;				/* sub one for ++ at top */
               i--;
	       if (*kf != lf)
	          (*++kf)->fld = buf + 1;	/* set for next FB_FLD */
               continue;
               }
            if (*buf == cdb_EOREC)
               break;
	    if (i > rlen)
	       return(FB_ERROR);
	    if (*buf == NULL){
	       /* skip any FB_FORMULA or FB_LINK fields */
	       for (; *kf != lf && 
	             ((*kf)->type == FB_FORMULA ||
                      (*kf)->dflink != NULL ||
                      (*kf)->type == FB_LINK) ; ){
	          p = (*kf)->fld;		/* save fld pointer in p*/
		  (*kf)->fld = buf;		/* make F point to NULL */
		  (*++kf)->fld = p;		/* make next FB_FLD use p */
		  }
	       if (*kf != lf)
	          (*++kf)->fld = buf + 1;	/* set for next FB_FLD */
	       }
	    }
	 return(i);
      }
